### ISSUE TITLE
Possible bugs in one-class SVM -- the parity of the number of data may break the performance.

### ISSUE THREAD

**Issue by b821213** [cjlin1/libsvm#69]
I am now using one-class SVM to learn a model from a dataset with 271 data.
All of the 271 data's label is +1, of course.
To measure the performance of a model, I first take a look at the accuracy of the results on training data -- it shouldn't be too bad, at least.

Here are the steps that I found and confirmed the problem:

I. I train a model by the first 270 data, the model I get correctly predicts that the last datum should be positive, and the accuracy of the results on training data seems good.

II. When I train a model by all these 271 data, the performance of the model on training data suddenly drops a lot, and it even predicts the last datum to be negative.

> A possible reason for this phenomenon may be that the last datum is overfitted, but it is hard to believe that a model derived from 270 data can be changed so much by merely a datum and that the overfitted datum is predicted to be negative.

III. To make the model even fits the last datum more, I train a model by 272 data -- the original 271 data with one copy of the last datum. And out of my expectation, the performance becomes good again.

> This doesn't make sense if the reason is overfitting. So I guess that the problem is on the parity of the size of the data. To test my guess here is step IV.

IV. To avoid the possibility that the last datum is weird, I use only the first 270 data to do this test. Every time I randomly choose _i_ data (_i_ = 0~9) and duplicate them to form a dataset with size 270 + _i_. Then I train a model on the dataset and see its performance. For each _i_, I will run 30 times and pick the average value as the result. The results obviously show that when _i_ is odd, the performance will be very awful, while this phenomenon never appears in the cases _i_ is even.

[This](http://codepad.org/WbFV9bJI) is the code of the 4 tests mentioned above. Though I use sklearn in the code, this problem can be reproduced by directly using libsvm as well (with gamma=1/n_features, which is the 'auto' value in sklearn). The dataset can be downloaded [here](https://drive.google.com/file/d/0B3AB2cDxZCwhNkUyMWo5WkV0U2M/view?usp=sharing). For now, a solution is that the user can always keep the size of the training dataset to be even.

**Comment by cjlin1** [comment]
Do you have the output of steps I and II?
